### ISSUE TITLE
Exception in socket initiator thread crashing the whole application

### ISSUE THREAD

**Issue by joseph-orbis** [connamara/quickfixn#257]
In our test environment we've encountered two occasions where an exception in the socket initiator thread crashed the whole application. Below are the two different exceptions causing the crashes. 

2014-03-04 03:12:11,255 [20] ERROR Orbis.Trading.Common.Services.BaseService System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'FileLog'.
   at QuickFix.FileLog.DisposedCheck()
   at QuickFix.FileLog.OnEvent(String s)
   at QuickFix.Transport.SocketInitiator.SocketInitiatorThreadStart(Object socketInitiatorThread)
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj

System.IO.IOException: Unable to write data to the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host
   at System.Net.Sockets.Socket.Send(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   at QuickFix.SocketInitiatorThread.Send(String data)
   at QuickFix.Session.Send(String message)
   at QuickFix.Session.SendRaw(Message message, Int32 seqNum)
   at QuickFix.Session.GenerateTestRequest(String id)
   at QuickFix.Session.Next()
   at QuickFix.SocketInitiatorThread.Read()

**Comment by AtanuSinha** [comment]
Any solution if you still wanna continue with FileLog

Here's my code snippet:

Thread v = new Thread(() => SendMsg(sessionID));
try
{
 v.Start();
}
catch(System.ObjectDisposedException oEx)
{
 //log this...
}

public void SendMsg(SessionID sessionID)
{
    using((.......initiate initiator here ....))
   {
      try
{
 initiator.start();

//Do stuff here.............................

catch()
{}
finally
{ initiator.stop()}
}
   }
}


it still crashes my main application. How can I avoid crashing the main application?

**Comment by andreimarculescu** [comment]
For those looking for a workaround, here is what I developed on my project:
- a custom ILogFactory implementation decorating FileLogFactory
- a custom ILog implementation decorating an arbitrary ILog implementation. Before any logging operation it checks if the Dispose method was called and, if so, it recreates the underlying ILog instance. In multithreaded context it might happen that Dispose is called after EnsureNotDisposed and before the logging action is executed on the underlying log. In this case I catch ObjectDisposedException and retry the log action after recreating the underlying log. Also note that the decorated instance is static so that multiple FileLog pointing to the same file are not created. Assumption is made SessionID doesn't change

ILogFactory implementation:
```cs
using QuickFix;

namespace PerseusLib.FixApi
{
    class SafeFileLogFactory : ILogFactory
    {
        private readonly FileLogFactory _quickFixFactory;

        public SafeFileLogFactory(SessionSettings settings)
        {
            _quickFixFactory = new FileLogFactory(settings);
        }

        public ILog Create(SessionID sessionId)
        {
            return new SafeFileLog(_quickFixFactory, sessionId);
        }
    }
}
```

ILog implementation
```cs
using System;
using QuickFix;

namespace PerseusLib.FixApi
{
    internal sealed class SafeFileLog : ILog
    {
        private readonly ILogFactory _quickFixFactory;
        private readonly SessionID _sessionId;
        private static volatile ILog _quickFixFileLog;
        private static volatile bool _shouldCreateFileLog = true;
        private static readonly object _creationLock = new object();

        public SafeFileLog(ILogFactory quickFixFactory, SessionID sessionId)
        {
            _quickFixFactory = quickFixFactory;
            _sessionId = sessionId;
        }

        public void Dispose()
        {
            if (!_shouldCreateFileLog)
            {
                lock (_creationLock)
                {
                    if (!_shouldCreateFileLog)
                    {
                        _shouldCreateFileLog = true;
                        _quickFixFileLog.Dispose();
                    }
                }
            }
        }

        public void Clear()
        {
            Action logAction = () => _quickFixFileLog.Clear();
            SafeLogAction(logAction);
        }

        public void OnIncoming(string msg)
        {
            Action logAction = () => _quickFixFileLog.OnIncoming(msg);
            SafeLogAction(logAction);
        }

        public void OnOutgoing(string msg)
        {
            Action logAction = () => _quickFixFileLog.OnOutgoing(msg);
            SafeLogAction(logAction);
        }

        public void OnEvent(string s)
        {
            Action logAction = () => _quickFixFileLog.OnEvent(s);
            SafeLogAction(logAction);
        }

        private void EnsureNotDisposed()
        {
            if (_shouldCreateFileLog)
            {
                lock (_creationLock)
                {
                    if (_shouldCreateFileLog)
                    {
                        RecreateUnderlyingLog();
                    }
                }
            }
        }

        private void RecreateUnderlyingLog()
        {
            _quickFixFileLog = _quickFixFactory.Create(_sessionId);
            _shouldCreateFileLog = false;
        }

        private void SafeLogAction(Action logAction)
        {
            EnsureNotDisposed();

            try
            {
                logAction();
            }
            catch (ObjectDisposedException)
            {
                // It might occur if Dispose is called avter the call to EnsureNotDisposed
                lock (_creationLock)
                {
                    RecreateUnderlyingLog();

                    logAction();
                }
            }
        }
    }
}

```

**Comment by francis2** [comment]
I agree about not throwing in FileLog. This seemingly insignificant change was introduced after the 1.5 release and it is not how quickfix.net worked so it caused an unhandled error in 7+ year old code.

**Comment by tr8dr** [comment]
I too am having this issue.  When I lose connection with an exchange am recreating the SocketInitiator, after having Stop()'ed the prior.   What is the preferred method to reinitiate the connection?

**Comment by francis2** [comment]
QF automatically reconnects after a connection loss. You do not need to recreate the SocketInitiator.

Whenever I force a disconnect, I call Disconnect(), then Stop() (be sure to trap exceptions), then GC.Collect() and GC.WaitForPendingFinalizers() (these last two were required with QF/C. I am not sure they're necessary with QF/net but I left them there as they don't hurt).

**Comment by hadagali** [comment]
Hi I am trying to use FIX50SP2_EP194.xml for my application Quickfix/N throws exception with error as 
BeginString=FIXT.1.1
DefaultApplVerID=9 is invalid 
TransportDataDictionary=C:\Projects\Shri\CommonFIX\spec\fix\FIXT11.xml
AppDataDictionary=C:\Projects\Shri\CommonFIX\spec\fix\FIX50SP2_EP194.xml


![image](https://cloud.githubusercontent.com/assets/11888222/9637143/285b65d0-516d-11e5-975f-d9cb056c9fd7.png)

**Comment by hadagali** [comment]
Any help greatly appreciated, thx

**Comment by gbirchmeier** [comment]
@hadagali - Github is not for help requests, and furthermore, what you've posted here is off-topic for this particular issue.  I'm deleting your comments.  I ask that you look at the support options listed here: http://quickfixn.org/help.html

**Comment by gbirchmeier** [comment]
addressed by #505
